 design/README.MD

```diff
@@ -0,0 +1,456 @@
+`() => "Hello World!"`
+generate interfaces that can be used like this:
+1. Dafny -> `var hello: seq<char> :- greet();`
+1. Java -> `String hello = greet();`
+1. C# -> `string hello = greet();`
+1. Python -> `hello = greet()`
+1. Rust -> `let hello: String = greet();`
+
+## Definitions
+
+- Runtime or native runtime:
+    A Dafny translation target language,
+    e.g Java or C#.
+    This also may include additional context
+    like language version or other versions/features.
+- Native type or native interface:
+    A type/interface in a runtime,
+    e.g. Java `String`
+- Smithy-runtime:
+    The project used to generate native types
+    for Smithy shapes.
+- Polymorph or polymorph-runtime:
+    The transformation layer between Dafny types
+    and native types,
+    e.g. polymorph-java.
+- Shim:
+    The object that exports the native interface
+    and connects to the internal Dafny implementation.
+- Type converters, ToDafny, ToNative:
+    The dual functions that convert between
+    Dafny and native types.
+    e.g. `D == ToDafny(ToNative(D))`,
+    `N == ToNative(ToDafny(N))`
+
+### Conventions used in this document
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL"
+in this document are to be interpreted as described in [RFC 2119](https://tools.ietf.org/html/rfc2119).
+
+## Smithy-Dafny modes
+
+### Library
+
+`smithy-dafny` supports `@localService` development.
+This is where the implementation is locally executing code.
+Rather than a remote service like most AWS Service Smithy models.
+
+In this mode `smithy-dafny` will produce Dafny types and interfaces
+for a Smithy model in Dafny.
+This creates an `abstract module` in Dafny you can refine
+with a concrete implementation.
+
+`smithy-dafny` also generates polymorph code
+that will expose native types and interfaces
+for your Smithy model.
+This creates a shim with native interfaces.
+These native interfaces are connected
+to the Dafny interface
+and takes care of type conversion back and forth.
+
+```mermaid
+---
+title: Smithy Dafny Library
+---
+flowchart RL
+subgraph Library[Native Library]
+    direction RL
+    Interface[Public Interface] --> ToDafny[Convert input \nto Dafny]
+    ToDafny --> DafnyImplementation[Dafny implementation]
+    DafnyImplementation --> ToNative[Convert output \nto native]
+    ToNative --> Interface
+end
+
+UserCode --> Interface
+Interface --> UserCode
+```
+
+The goal is for native developers
+to call Dafny libraries with idiomatic types.
+
+### Wrapped AWS Services
+
+`smithy-dafny` also supports wrapping existing AWS Services
+with existing implementations that have a native implementation.
+
+The problem we are trying to solve here is:
+I am writing code in Dafny
+and I want to call an AWS service,
+how do I do that?
+
+In this mode `smithy-dafny` will produce Dafny types and interfaces
+so that Dafny code can call the native implementation.
```

Comment:
isn't client generation something which Smithy does generally? i'm unclear on how this would work. "wrapping a client" sounds like runtime behavior, but Dafny doesn't have a runtime, right? e.g. if Java is a transpilation target for Dafny and `smithy-dafny` is configured to wrap the native Java client in order to make calls to some AWS service, then what happens for non-Java transpilation targets?

Reply:
Each transpilation targets would need to have its own way to provide a native SDK to be wrapped.